experimental/ActionPolyRing/src/Types.jl

```diff
@@ -0,0 +1,203 @@
+###############################################################################
+#
+#  Action polynomial rings 
+#
+###############################################################################
+
+abstract type ActionPolyRing{T} <: Ring end 
+
+abstract type ActionPolyRingElem{T} <: RingElem end
+
+### Difference ###
+mutable struct DifferencePolyRing{T} <: ActionPolyRing{T}
+  upoly_ring::AbstractAlgebra.Generic.UniversalPolyRing{T}
+  elementary_symbols::Vector{Symbol}
+  ndiffs::Int
+  are_perms_up_to_date::Bool
+  jet_to_var::Any #Always of type Dict{Tuple{Int, Vector{Int}}, DifferencePolyRingElem{T}}
+  var_to_jet::Any #Always of type Dict{DifferencePolyRingElem{T}, Tuple{Int, Vector{Int}}}
+  jet_to_upoly_idx::Dict{Tuple{Int, Vector{Int}}, Int}
+  ranking::Any #Alyways of type DifferenceRanking{T}
+  permutation::Vector{Int}
+
+  function DifferencePolyRing{T}(R::Ring, nelementary_symbols::Int, ndiffs::Int) where {T}
+    @req nelementary_symbols >= 1 "The number of elementary symbols must be positive"
+    @req ndiffs >= 1 "The number of endomorphisms must be positive"
+    elementary_symbols = map(x -> Symbol("u" * string(x)), 1:nelementary_symbols)
+    upoly_ring = universal_polynomial_ring(R; cached = false)
+    
+    jet_to_var = Dict{Tuple{Int, Vector{Int}}, DifferencePolyRingElem{T}}()
+    var_to_jet = Dict{DifferencePolyRingElem{T}, Tuple{Int, Vector{Int}}}()
+    jet_to_upoly_idx = Dict{Tuple{Int, Vector{Int}}, Int}()
+    
+    return new{T}(upoly_ring, elementary_symbols, ndiffs, false, jet_to_var, var_to_jet, jet_to_upoly_idx)
+  end
```

Comment:
to avoid code duplication, you can just call one constructor from the other one like this:
```suggestion
  function DifferencePolyRing{T}(R::Ring, nelementary_symbols::Int, ndiffs::Int) where {T}
    @req nelementary_symbols >= 1 "The number of elementary symbols must be positive"
    elementary_symbols = map(x -> Symbol('u', x), 1:nelementary_symbols)
        
    return DifferencePolyRing{T}(R, elementary_symbols, ndiffs)
  end
```

Alternatively, you could completely remove this inner constructor, and put the symbol generation into the `difference_polynomial_ring` function